### PR TITLE
fix(ssr): vue-ssr-webpack-plugin compatible with webpack 4/5

### LINKED ISSUE #11718: vuejs/vue-ssr-webpack-plugin and webpack 5
### What problem does this feature solve?
github missing repository vuejs/vue-ssr-webpack-plugin, i update webpack to version 5 and have error
[vue-server-renderer-webpack-plugin] webpack config `output.libraryTarget` should be "commonjs2".

### What does the proposed API look like?
in my output.libraryTarget is libraryTarget: 'commonjs2'

### PR DESCRIPTION
**Pull request by ygj6**
close #11718

**What kind of change does this PR introduce?** (check at least one)

- [x] Bugfix
- [ ] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch for v2.x (or to a previous version branch), _not_ the `master` branch
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] All tests are passing: https://github.com/vuejs/vue/blob/dev/.github/CONTRIBUTING.md#development-setup
- [ ] New/updated tests are included

If adding a **new feature**, the PR's description includes:
- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**

### CODE REVIEW

**Comment by julian91** [comment]
+1 really need this

**Comment by vsetka** [comment]
Can this be merged?

File: src/server/webpack-plugin/util.js [hunk 1]
```diff
@@ -9,8 +9,15 @@ export const validate = compiler => {
...
-  if (compiler.options.output && compiler.options.output.libraryTarget !== 'commonjs2') {
-    warn('webpack config `output.libraryTarget` should be "commonjs2".')
+  if (compiler.options.output) {
+    // Webpack < 5.0.0
+    if (compiler.options.output.libraryTarget && compiler.options.output.libraryTarget !== 'commonjs2') {
+      warn('webpack config `output.libraryTarget` should be "commonjs2".')
+    }
+    // Webpack >= 5.0.0
+    else if (compiler.options.output.library && compiler.options.output.library.type !== 'commonjs2') {
+      warn('webpack config `output.libraryTarget` should be "commonjs2".')
```
Comment (blake-newman) [inline src/server/webpack-plugin/util.js:19]: `output.library.type`

**Review by blake-newman — CHANGES_REQUESTED** [review]

**Comment by yulodl** [comment]
Waiting for update

**Comment by beckybair** [comment]
what about making these changes on vue-server-renderer/server-plugin.js?
https://github.com/vuejs/vue/issues/11718#issuecomment-717786088

File: src/server/webpack-plugin/util.js (hunk 1 above)
Comment (tibineagu) [reply src/server/webpack-plugin/util.js:19]: @blake-newman According to the [webpack 5 documentation](https://webpack.js.org/configuration/output/#outputlibrary), I dont think `output.library.type` is what's required.

The [config supports `output.libraryTarget: 'commonjs2'`](https://webpack.js.org/configuration/output/#librarytarget-commonjs2) as until now.

File: src/server/webpack-plugin/client.js [hunk 2]
```diff
@@ -19,6 +20,13 @@ export default class VueSSRClientPlugin {
       const initialFiles = uniq(Object.keys(stats.entrypoints)
         .map(name => stats.entrypoints[name].assets)
         .reduce((assets, all) => all.concat(assets), [])
+        .map(file => {
+          if (isObject(file) && file.name) {
+            return file.name;
+          } else {
+            return file;
```
Comment (lmiller1990) [inline src/server/webpack-plugin/client.js:27]:
```suggestion
          if (isObject(file) && file.name) {
            return file.name
          } else {
            return file
```

File: src/server/webpack-plugin/server.js [hunk 3]
```diff
@@ -20,7 +21,14 @@ export default class VueSSRServerPlugin {
         return cb()
       }
 
-      const entryAssets = entryInfo.assets.filter(isJS)
+      const entryAssets = entryInfo.assets
+        .map(file => {
+          if ( isObject(file) && file.name) {
+            return file.name;
+          } else {
+            return file;
```
Comment (lmiller1990) [inline src/server/webpack-plugin/server.js:29]:
```suggestion
          if ( isObject(file) && file.name) {
            return file.name
          } else {
            return file
```

File: src/server/webpack-plugin/server.js [hunk 4]
```diff
@@ -45,6 +53,9 @@ export default class VueSSRServerPlugin {
       stats.assets.forEach(asset => {
         if (isJS(asset.name)) {
           bundle.files[asset.name] = compilation.assets[asset.name].source()
+          if (asset.info && asset.info.related && asset.info.related.sourceMap) {
+            bundle.maps[asset.info.related.sourceMap.replace(/\.map$/, '')] = JSON.parse(compilation.assets[asset.info.related.sourceMap].source());
```
Comment (lmiller1990) [inline src/server/webpack-plugin/server.js:57]:
```suggestion
            bundle.maps[asset.info.related.sourceMap.replace(/\.map$/, '')] = JSON.parse(compilation.assets[asset.info.related.sourceMap].source())
```

File: src/server/template-renderer/index.js [hunk 5]
```diff
@@ -224,7 +224,7 @@ export default class TemplateRenderer {
     if (this.clientManifest) {
       const initial = this.preloadFiles.filter(({ file }) => isJS(file))
       const async = (this.getUsedAsyncFiles(context) || []).filter(({ file }) => isJS(file))
-      const needed = [initial[0]].concat(async, initial.slice(1))
+      const needed = [].concat(initial[0] ? [initial[0]] : [], async, initial.slice(1));
```
Comment (posva) [inline src/server/template-renderer/index.js:227]:
```suggestion
      const needed = [].concat(initial[0] ? [initial[0]] : [], async, initial.slice(1))
```

**Comment by redneckz** [comment]
Hi colleagues.
Have a nice day.

The last activity was several weeks ago...
Any progress on this issue?

Currently, we try to move to Webpack 5 and have troubles with Vue SSR plugin.
We stuck trying to migrate to micro frontend architecture (by means of Module Federation). In our case, it is a blocker :(

Please give some roadmap to adjust our deadlines and make appropriate decisions.

Thanks in advance.

**Pull request closed without merging**